ml-algorithms/src/main/java/org/opensearch/ml/engine/tools/SearchIndexTool.java

```diff
@@ -0,0 +1,195 @@
+import org.opensearch.core.xcontent.XContentParser;
+import org.opensearch.ml.common.spi.tools.Tool;
+import org.opensearch.ml.common.spi.tools.ToolAnnotation;
+import org.opensearch.ml.common.transport.connector.MLConnectorSearchAction;
+import org.opensearch.ml.common.transport.model.MLModelSearchAction;
+import org.opensearch.search.SearchHit;
+import org.opensearch.search.builder.SearchSourceBuilder;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonObject;
+
+import lombok.Getter;
+import lombok.Setter;
+import lombok.extern.log4j.Log4j2;
+
+@Getter
+@Setter
+@Log4j2
+@ToolAnnotation(SearchIndexTool.TYPE)
+public class SearchIndexTool implements Tool {
+
+    public static final String INPUT_FIELD = "input";
+    public static final String INDEX_FIELD = "index";
+    public static final String QUERY_FIELD = "query";
+
+    public static final String TYPE = "SearchIndexTool";
+    private static final String DEFAULT_DESCRIPTION =
+        "Use this tool to search index with a query. You should pass in two parameters: index and query. Index is the index name and the query is an OpenSearch DSL query.";
+
+    private String name = TYPE;
+
+    private String description = DEFAULT_DESCRIPTION;
+
+    private Client client;
+
+    private NamedXContentRegistry xContentRegistry;
+
+    public static final Gson gson;
+
+    static {
+        gson = new Gson();
+    }
+
+    public SearchIndexTool(Client client, NamedXContentRegistry xContentRegistry) {
+        this.client = client;
+        this.xContentRegistry = xContentRegistry;
+    }
+
+    @Override
+    public String getType() {
+        return TYPE;
+    }
+
+    @Override
+    public String getVersion() {
+        return null;
+    }
+
+    @Override
+    public boolean validate(Map<String, String> parameters) {
+        if (parameters == null || !parameters.containsKey(INPUT_FIELD)) {
+            return false;
+        }
+        String input = parameters.get(INPUT_FIELD);
+        return input != null;
+    }
+
+    @Override
+    public <T> void run(Map<String, String> parameters, ActionListener<T> listener) {
+        try {
+            String input = parameters.get(INPUT_FIELD);
+            String index = "";
+            String query = "";
+            try {
+                // need this try catch block, because index, query field may not exist
+                JsonObject jsonObject = gson.fromJson(input, JsonObject.class);
+                index = jsonObject.get(INDEX_FIELD).getAsString();
+                query = jsonObject.get(QUERY_FIELD).toString();
+                query = "{\"query\": " + query + "}";
+            } catch (Exception e) {
+                // throw new IllegalArgumentException("wrong input");
+            }
+            if (org.apache.commons.lang3.StringUtils.isBlank(query)) {
+                throw new IllegalArgumentException("[" + QUERY_FIELD + "] is null or empty, can not process it.");
```

Comment:
If a listener is shown, you should use `listener.onFailure(throw new IllegalArgumentException("[" + QUERY_FIELD + "] is null or empty, can not process it."));` because I remember agent framework doesn't catch exceptions.

Reply:
Ok

Reply:
Since query will be wrapped like `query = "{\"query\": " + query + "}";`, query will never be blank. Therefore, there is no need to catch this exception.